dcos/util.go

```diff
@@ -0,0 +1,297 @@
+	}
+
+	if d.roleAgentPublicFile == "" {
+		d.roleAgentPublicFile = FileAgentPublic
+	}
+
+	// setup mesos state URL
+	if d.mesosStateURL == "" {
+		stateURL := url.URL{
+			Scheme: "https",
+			Host:   DNSRecordLeader,
+			Path:   "/mesos/state",
+		}
+		d.mesosStateURL = stateURL.String()
+	}
+
+	return d, nil
+}
+
+// DetectIP returns an output from `FileDetectIP` script.
+// This is a nice way of shelling out to `detect_ip` script which handles timeout.
+func (d *dcosUtil) DetectIP() (net.IP, error) {
+	if _, err := os.Stat(d.detectIPLocation); err != nil {
+		return nil, err
+	}
+
+	ce, err := exec.Run("sh", []string{d.detectIPLocation}, exec.Timeout(time.Millisecond*100))
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := ioutil.ReadAll(ce)
+	if err != nil {
+		return nil, err
+	}
+
+	err = <-ce.Done
+	if err != nil {
+		return nil, err
+	}
+
+	// strip the trailing \n
+	detectedIP := strings.TrimRight(string(buf), "\n")
+	if detectedIP == "" {
+		return nil, ErrDCOSUtil{fmt.Sprintf("command %s return empty output", FileDetectIP)}
+	}
+
+	validIP := net.ParseIP(detectedIP)
+	if validIP == nil {
+		return nil, ErrDCOSUtil{fmt.Sprintf("command %s return invalid IP address %s", FileDetectIP, detectedIP)}
+	}
+
+	return validIP, nil
+}
+
+// Role return a node role.
+func (d *dcosUtil) Role() (string, error) {
+	var isMaster, isAgent, isAgentPublic bool
+	isFile := func(path string, flag *bool) {
+		if _, err := os.Stat(path); err == nil {
+			*flag = true
+		}
+	}
+
+	// iterate over files and get the role.
+	for _, i := range []struct {
+		file string
+		flag *bool
+	}{
+		{
+			file: d.roleMasterFile,
+			flag: &isMaster,
+		},
+		{
+			file: d.roleAgentFile,
+			flag: &isAgent,
+		},
+		{
+			file: d.roleAgentPublicFile,
+			flag: &isAgentPublic,
+		},
+	} {
+		isFile(i.file, i.flag)
+	}
+
+	// make sure we have only one role
+	if isMaster && (isAgent || isAgentPublic) {
+		return "", ErrDCOSUtil{"Node cannot be master and agent / agent_public at the same time"}
+	}
+
+	if isAgent && isAgentPublic {
+		return "", ErrDCOSUtil{"Node cannot be agent and agent_public at the same time"}
+	}
+
+	if !isMaster && !isAgent && !isAgentPublic {
+		return "", ErrDCOSUtil{"Unable to determine node role"}
+	}
+
+	// iterate over roles
+	for _, nodeRole := range []struct {
+		flag bool
+		role string
+	}{
+		{
+			flag: isMaster,
+			role: RoleMaster,
+		},
+		{
+			flag: isAgent,
+			role: RoleAgent,
+		},
+		{
+			flag: isAgentPublic,
+			role: RoleAgentPublic,
+		},
+	} {
+		if nodeRole.flag {
+			return nodeRole.role, nil
+		}
+	}
+
+	return "", ErrDCOSUtil{"Unable to determine node role"}
+}
+
+// IsLeader checks if the node is leader.
+func (d *dcosUtil) IsLeader() (bool, error) {
+	role, err := d.Role()
+	if err != nil {
+		return false, err
+	}
+
+	// agent cannot be leader
+	if role != RoleMaster {
+		return false, nil
+	}
+
+	addrs, err := net.LookupIP(DNSRecordLeader)
+	if err != nil {
+		return false, err
+	}
+
+	if len(addrs) != 1 {
```

Comment:
won't this blow up as soon as we support both ipv4 and ipv6? why not just add these as keys of a map and check if the result of `DetectIP` is in the key set?